// @polkadot/metadata/index.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Metadata", {
  enumerable: true,
  get: function () {
    return _Metadata.Metadata;
  }
});
Object.defineProperty(exports, "decorateConstants", {
  enumerable: true,
  get: function () {
    return _index.decorateConstants;
  }
});
Object.defineProperty(exports, "decorateExtrinsics", {
  enumerable: true,
  get: function () {
    return _index.decorateExtrinsics;
  }
});
Object.defineProperty(exports, "decorateStorage", {
  enumerable: true,
  get: function () {
    return _index.decorateStorage;
  }
});
Object.defineProperty(exports, "expandMetadata", {
  enumerable: true,
  get: function () {
    return _index.expandMetadata;
  }
});
Object.defineProperty(exports, "packageInfo", {
  enumerable: true,
  get: function () {
    return _packageInfo.packageInfo;
  }
});

require("./detectPackage.cjs");

var _Metadata = require("./Metadata.cjs");

var _index = require("./decorate/index.cjs");

var _packageInfo = require("./packageInfo.cjs");
