// @polkadot/api/index.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Keyring", {
  enumerable: true,
  get: function () {
    return _keyring.Keyring;
  }
});
Object.defineProperty(exports, "WsProvider", {
  enumerable: true,
  get: function () {
    return _ws.WsProvider;
  }
});
Object.defineProperty(exports, "packageInfo", {
  enumerable: true,
  get: function () {
    return _packageInfo.packageInfo;
  }
});
Object.defineProperty(exports, "ApiPromise", {
  enumerable: true,
  get: function () {
    return _index.ApiPromise;
  }
});
Object.defineProperty(exports, "SubmittableResult", {
  enumerable: true,
  get: function () {
    return _index2.SubmittableResult;
  }
});
Object.defineProperty(exports, "ApiRx", {
  enumerable: true,
  get: function () {
    return _index3.ApiRx;
  }
});

require("./detectPackage.cjs");

var _keyring = require("@polkadot/keyring");

var _ws = require("@polkadot/rpc-provider/ws");

var _packageInfo = require("./packageInfo.cjs");

var _index = require("./promise/index.cjs");

var _index2 = require("./submittable/index.cjs");

var _index3 = require("./rx/index.cjs");
